File: 0440.K-thSmallestinLexicographicalOrder.go
func findKthNumber(n int, k int) int {
    cur, k := 1, k-1
    for k > 0 {
        step := count(n, cur, cur+1)
        if step <= k {
            cur++
            k -= step
        } else {
            cur *= 10
            k--
        }
    }
    return cur
}

func count(n, n1, n2 int) int {
    cnt := 0
    for n1 <= n {
        cnt += min(n+1, n2) - n1
        n1, n2 = 10*n1, 10*n2
    }
    return cnt
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

File: 0000-0999/0440.K-thSmallestinLexicographicalOrder.py
class Solution(object):
    def findKthNumber(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: int
        """
        ans = 1
        k -= 1
        while k > 0:
            count = 0
            interval = [ans, ans+1]
            while interval[0] <= n:
                count += min(n+1, interval[1]) - interval[0]
                interval = [interval[0]*10, interval[1]*10]
            if k >= count:
                ans += 1
                k -= count
            else:
                ans *= 10
                k -= 1
        return ans
